src/schemes/boundary/open_boundary/system.jl

```diff
@@ -0,0 +1,508 @@
+        rho_ref = reference_density(neighbor_position, t)
+
+        p = particle_pressure(v_neighbor_system, neighbor_system, neighbor)
+        p_ref = reference_pressure(neighbor_position, t)
+
+        v_neighbor = current_velocity(v_neighbor_system, neighbor_system, neighbor)
+        v_neighbor_ref = reference_velocity(neighbor_position, t)
+
+        # Determine characteristic variables
+        density_term = -sound_speed^2 * (rho - rho_ref)
+        pressure_term = p - p_ref
+        velocity_term = rho * sound_speed *
+                        (dot(v_neighbor - v_neighbor_ref, flow_direction))
+
+        kernel_ = smoothing_kernel(neighbor_system, distance)
+
+        characteristics[1, particle] += (density_term + pressure_term) * kernel_
+        characteristics[2, particle] += (velocity_term + pressure_term) * kernel_
+        characteristics[3, particle] += (-velocity_term + pressure_term) * kernel_
+
+        volume[particle] += kernel_
+    end
+
+    return system
+end
+
+@inline function prescribe_conditions!(characteristics, particle, ::OutFlow)
+    # J3 is prescribed (i.e. determined from the exterior of the domain).
+    # J1 and J2 is transimtted from the domain interior.
+    characteristics[3, particle] = zero(eltype(characteristics))
+
+    return characteristics
+end
+
+@inline function prescribe_conditions!(characteristics, particle, ::InFlow)
+    # Allow only J3 to propagate upstream to the boundary
+    characteristics[1, particle] = zero(eltype(characteristics))
+    characteristics[2, particle] = zero(eltype(characteristics))
+
+    return characteristics
+end
+
+@inline function update_quantities!(system::OpenBoundarySPHSystem, v, u, t)
+    (; density, pressure, characteristics, flow_direction, sound_speed,
+    reference_velocity, reference_pressure, reference_density) = system
+
+    @threaded for particle in each_moving_particle(system)
+        particle_position = current_coords(u, system, particle)
+
+        J1 = characteristics[1, particle]
+        J2 = characteristics[2, particle]
+        J3 = characteristics[3, particle]
+
+        rho_ref = reference_density(particle_position, t)
+        density[particle] = rho_ref + ((-J1 + 0.5 * (J2 + J3)) / sound_speed^2)
+
+        p_ref = reference_pressure(particle_position, t)
+        pressure[particle] = p_ref + 0.5 * (J2 + J3)
+
+        v_ref = reference_velocity(particle_position, t)
+        rho = density[particle]
+        v_ = v_ref + ((J2 - J3) / (2 * sound_speed * rho)) * flow_direction
+
+        for dim in 1:ndims(system)
+            v[dim, particle] = v_[dim]
+        end
+    end
+
+    return system
+end
+
+function check_domain!(system, v, u, v_ode, u_ode, semi)
+    # TODO: Is a thread supported version possible?
+    for particle in each_moving_particle(system)
+        foreach_system(semi) do fluid_system
+            check_fluid_domain!(system, fluid_system, particle, v, u, v_ode, u_ode, semi)
+        end
```

Comment:
What if you have multiple fluid systems? Will a particle leaving the inflow zone be spawned in each of the fluid systems?

Reply:
Good point! See [4aa2a89](https://github.com/trixi-framework/TrixiParticles.jl/pull/442/commits/4aa2a899eb2fa8b9dd1d2f48408363d922457c4a)

Reply:
Would it maybe make more sense to permanently link the open boundary system to one fluid system?

Reply:
I originally did it that way. But it was really cumbersome and didn't entail any advantages. The current approach is more TrixiParticles style, since we only need one open boundary for multiple fluid systems.

Reply:
But how do you use one open boundary for multiple fluid systems? Say one is using density diffusion and the other isn't, so you can compare them. So you have two pipe flows, one with and one without DD. How do I connect the no-DD open boundary to the no-DD fluid system and vice versa?

Reply:
I want particles coming out of one inflow to go into one fluid system and particles coming out of the other inflow to go into another fluid system.